lib/dry/validation/extensions/predicates_as_macros.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require 'dry/logic/predicates'
+require 'dry/schema/predicate_registry'
+require 'dry/validation/contract'
+
+module Dry
+  module Validation
+    # Encapsulation of a dry-logic predicate.
+    class Predicate
+      # List of predicates to be imported.
+      #
+      # @see Dry::Validation::Contract
+      WHITELIST = %i[
+        filled? gt? gteq? included_in? includes? inclusion? is? lt?
+        lteq? max_size? min_size? not_eql? odd? respond_to? size? true?
+        uuid_v4?
+      ].freeze
+
+      # @api private
+      REGISTRY = Dry::Schema::PredicateRegistry.new(
+        Dry::Logic::Predicates
+      ).freeze
+
+      # @api private
+      attr_reader :name
+
+      # @api private
+      def initialize(name)
+        @name = name
+      end
+
+      # @api private
+      def arg_names
+        REGISTRY.arg_list(name).map(&:first)
+      end
+
+      # @api private
+      def call(args)
+        REGISTRY[name].(*args)
+      end
+
+      # @api private
+      def add_failure_from_call(key, arg_values)
+        message_opts = arg_names.zip(arg_values).to_h
+
+        key.failure(name, message_opts)
+      end
```

Comment:
it would be better to just return `message_opts` here and rename this method to `message_opts` and then in `import_predicates_as_macros` you could just do `key.failure(name, predicate.message_opts)`. Otherwise, `Predicate` will be coupled to evaluator's key API.

Reply:
Yeah, much better. Done in the last commit.